File: frontend/lib/mappers.ts
import {
  AIReviewUpdateResponse,
  BackendAIReviewUpdateResponse,
  BackendCluster,
  BackendHomePageData,
  BackendSource,
  BackendStory,
  CompareCluster,
  HomePageData,
  PaginatedResponse,
  Source,
  Story,
} from './types'
import { AppLanguage } from './i18n'

function getCategoryImage(category?: string | null): string {
  const placeholders: Record<string, string> = {
    politics: '/images/politics-placeholder.svg',
    security: '/images/security-placeholder.svg',
    economy: '/images/economy-placeholder.svg',
    humanitarian: '/images/humanitarian-placeholder.svg',
    diaspora: '/images/diaspora-placeholder.svg',
    default: '/images/default-placeholder.svg',
  }

  const normalized = (category || '').toLowerCase()
  return placeholders[normalized] || placeholders.default
}

function mapFramingTone(
  tone?: string | null
): 'positive' | 'neutral' | 'negative' | 'mixed' {
  if (tone === 'positive' || tone === 'neutral' || tone === 'negative' || tone === 'mixed') {
    return tone
  }
  return 'neutral'
}

function detectStoryLanguage(story: BackendStory): AppLanguage {
  const sourceLanguage = story.source.language?.toLowerCase()

  if (sourceLanguage?.startsWith('so')) {
    return 'so'
  }

  if (sourceLanguage?.startsWith('en')) {
    return 'en'
  }

  const combinedText = `${story.title} ${story.excerpt || ''} ${story.summary || ''}`.toLowerCase()
  const somaliSignals = [
    'ayaa',
    'soomaaliya',
    'muqdisho',
    'wasaaradda',
    'wasiir',
    'dadka',
    'gobolka',
    'ciidamada',
    'warbixin',
    'maanta',
    'kadib',
    'qorshe',
    'cusub',
  ]

  const englishSignals = [
    'the ',
    'somalia',
    'government',
    'minister',
    'report',
    'president',
    'world',
    'after',
    'security',
    'economy',
  ]

  const somaliMatches = somaliSignals.filter((signal) => combinedText.includes(signal)).length
  const englishMatches = englishSignals.filter((signal) => combinedText.includes(signal)).length

  if (somaliMatches > englishMatches) {
    return 'so'
  }

  return 'en'
}

export function mapSource(source: BackendSource): Source {
  return {
    id: source.id,
    name: source.name,
    url: source.base_url || source.feed_url || '',
    category: source.category,
    description: source.description,
    language: source.language || undefined,
    country: source.country || undefined,
  }
}

export function mapStory(story: BackendStory): Story {
  return {
    id: story.id,
    slug: story.slug,
    lang: detectStoryLanguage(story),
    title: story.title,
    excerpt: story.excerpt || story.summary || '',
    content: story.content_html || undefined,
    summary: story.summary || undefined,
    translations: story.translations || undefined,
    source: mapSource(story.source),
    publishedAt: story.published_at,
    updatedAt: story.updated_at || undefined,
    region: story.region,
    category: story.category,
    topics: story.topics || [],
    framing: story.framing
      ? {
          id: story.framing.id,
          label: story.framing.label,
          description: story.framing.description,
          tone: mapFramingTone(story.framing.tone),
        }
      : undefined,
    imageUrl: story.image_url || getCategoryImage(story.category),
    originalUrl: story.original_url,
    readingTime: story.reading_time || undefined,
    clusterId: story.cluster_id || undefined,
    isBreaking: story.is_breaking,
  }
}

export function mapCluster(cluster: BackendCluster): CompareCluster {
  return {
    id: cluster.id,
    title: cluster.title,
    commonFacts: cluster.common_facts || '',
    coverageDifferences:
      cluster.coverage_differences || 'Coverage differences are still being enriched for this story.',
    neutralSummary: cluster.neutral_summary || '',
    keyThemes: cluster.key_themes || [],
    consensusLevel: cluster.consensus_level,
    aiNeutralSummary: cluster.ai_neutral_summary || undefined,
    aiCoverageDifferences: cluster.ai_coverage_differences || undefined,
    aiConsensusLevel: cluster.ai_consensus_level || undefined,
    aiKeyThemes: cluster.ai_key_themes || [],
    aiGeneratedAt: cluster.ai_generated_at || undefined,
    aiModelUsed: cluster.ai_model_used || undefined,
    hasAISynthesis: cluster.has_ai_synthesis || false,
    aiReviewStatus: cluster.ai_review_status || 'unreviewed',
    aiReviewNote: cluster.ai_review_note || undefined,
    aiReviewedAt: cluster.ai_reviewed_at || undefined,
    createdAt: cluster.created_at || undefined,
    updatedAt: cluster.updated_at || undefined,
    storyCount: cluster.story_count || undefined,
    stories: cluster.stories.map(mapStory),
    sources: cluster.sources.map(mapSource),
  }
}

export function mapHomePageData(home: BackendHomePageData): HomePageData {
  return {
    heroStory: mapStory(home.hero_story),
    secondaryStories: home.secondary_stories.map(mapStory),
    comparePreview: home.compare_preview ? mapCluster(home.compare_preview) : null,
    latestStories: home.latest_stories.map(mapStory),
    somaliaStories: home.somalia_stories.map(mapStory),
    worldStories: home.world_stories.map(mapStory),
  }
}

export function mapPaginatedStories(
  response: PaginatedResponse<BackendStory>
): PaginatedResponse<Story> {
  return {
    items: response.items.map(mapStory),
    total: response.total,
    limit: response.limit,
    offset: response.offset,
  }
}

export function mapPaginatedClusters(
  response: PaginatedResponse<BackendCluster>
): PaginatedResponse<CompareCluster> {
  return {
    items: response.items.map(mapCluster),
    total: response.total,
    limit: response.limit,
    offset: response.offset,
  }
}

export function mapAIReviewUpdateResponse(
  response: BackendAIReviewUpdateResponse
): AIReviewUpdateResponse {
  return {
    clusterId: response.cluster_id,
    aiReviewStatus: response.ai_review_status,
    aiReviewNote: response.ai_review_note || undefined,
    aiReviewedAt: response.ai_reviewed_at || undefined,
  }
}

File: backend/app/mappers.py
from __future__ import annotations

import json

from app import models, schemas


def map_source_to_response(db_source: models.Source) -> schemas.Source:
    return schemas.Source.model_validate(db_source)


def map_story_to_response(db_story: models.Story) -> schemas.Story:
    framing = None
    if db_story.framing_label:
        framing = schemas.Framing(
            id=f"{db_story.id}-framing",
            label=db_story.framing_label,
            description=db_story.framing_description,
            tone=db_story.framing_tone,
        )

    return schemas.Story(
        id=db_story.id,
        slug=db_story.slug,
        title=db_story.title,
        excerpt=db_story.excerpt,
        content_html=db_story.content_html,
        summary=db_story.summary,
        source_id=db_story.source_id,
        original_url=db_story.original_url,
        canonical_url_hash=db_story.canonical_url_hash,
        published_at=db_story.published_at,
        updated_at=db_story.updated_at,
        fetched_at=db_story.fetched_at,
        region=db_story.region,
        category=db_story.category,
        topics=list(db_story.topics or []),
        translations=dict(db_story.translations or {}),
        image_url=db_story.image_url,
        reading_time=db_story.reading_time,
        is_breaking=db_story.is_breaking,
        cluster_id=db_story.cluster_id,
        source=map_source_to_response(db_story.source),
        framing=framing,
        created_at=db_story.created_at,
    )


def _coerce_theme_list(raw_value) -> list[str]:
    if raw_value is None:
        return []
    if isinstance(raw_value, list):
        return [str(item).strip() for item in raw_value if str(item).strip()]
    if isinstance(raw_value, str):
        cleaned = raw_value.strip()
        if not cleaned:
            return []
        try:
            parsed = json.loads(cleaned)
        except json.JSONDecodeError:
            return []
        if isinstance(parsed, list):
            return [str(item).strip() for item in parsed if str(item).strip()]
    return []


def map_cluster_to_response(db_cluster: models.Cluster) -> schemas.Cluster:
    story_items = [map_story_to_response(story) for story in db_cluster.stories]
    unique_sources: dict[str, schemas.Source] = {}
    for story in db_cluster.stories:
        unique_sources[story.source.id] = map_source_to_response(story.source)

    return schemas.Cluster(
        id=db_cluster.id,
        title=db_cluster.title,
        common_facts=db_cluster.common_facts,
        coverage_differences=db_cluster.coverage_differences,
        neutral_summary=db_cluster.neutral_summary,
        key_themes=list(db_cluster.key_themes or []),
        consensus_level=db_cluster.consensus_level,
        ai_neutral_summary=db_cluster.ai_neutral_summary,
        ai_coverage_differences=db_cluster.ai_coverage_differences,
        ai_consensus_level=db_cluster.ai_consensus_level,
        ai_key_themes=_coerce_theme_list(db_cluster.ai_key_themes),
        ai_generated_at=db_cluster.ai_generated_at,
        ai_model_used=db_cluster.ai_model_used,
        has_ai_synthesis=db_cluster.has_ai_synthesis,
        ai_review_status=db_cluster.ai_review_status or "unreviewed",
        ai_review_note=db_cluster.ai_review_note,
        ai_reviewed_at=db_cluster.ai_reviewed_at,
        story_count=db_cluster.story_count,
        created_at=db_cluster.created_at,
        updated_at=db_cluster.updated_at,
        stories=story_items,
        sources=list(unique_sources.values()),
    )
